component/remote/vault/vault.go

```diff
@@ -0,0 +1,237 @@
+	}
+	return a.Auth[0].authMethod()
+}
+
+func (a *Arguments) secretStore(cli *vault.Client) secretStore {
+	// TODO(rfratto): support different stores (like a logical store).
+	return &kvStore{c: cli}
+}
+
+// ClientOptions sets extra options on the Client.
+type ClientOptions struct {
+	MinRetryWait time.Duration `river:"min_retry_wait,attr,optional"`
+	MaxRetryWait time.Duration `river:"max_retry_wait,attr,optional"`
+	MaxRetries   int           `river:"max_retries,attr,optional"`
+	Timeout      time.Duration `river:"timeout,attr,optional"`
+}
+
+// Exports is the values exported by remote.vault.
+type Exports struct {
+	// Data holds key-value pairs returned from Vault after retrieving the key.
+	// Any keys-value pairs returned from Vault which are not []byte or strings
+	// cannot be represented as secrets and are therefore ignored.
+	//
+	// However, it seems that most secrets engines don't actually return
+	// arbitrary data, so this limitation shouldn't cause any issues in practice.
+	Data map[string]rivertypes.Secret `river:"data,attr"`
+}
+
+// Component implements the remote.vault component.
+type Component struct {
+	opts component.Options
+	log  log.Logger
+
+	// stateChangedCh is written to when client, authToken, or secret are
+	// recreated when Updating. This triggers the Run goroutine to start a new
+	// loop for renewal.
+	//
+	// stateChangedCh is not written to if Run directly changes the
+	// authToken/secret fields.
+	stateChangedCh chan struct{}
+
+	// mut is used to protect the client, authToken, and secret fields to be set
+	// to new pointers.
+	//
+	// The contents of the fields must be considered read only, as the pointer
+	// values are passed along to Vault APIs which have no knowledge of the
+	// mutex.
+	mut       sync.RWMutex
+	client    *vault.Client // Current client used to make requests to Vault.
+	authToken *vault.Secret // Secret used to authenticate to vault (nil for auth.token)
+	secret    *vault.Secret // Secret retrieved from vault
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+	_ component.DebugComponent  = (*Component)(nil)
+)
+
+// New creates a new remote.vault component. It will try to immediately read
+// the secret from Vault and return an error if the secret can't be read or if
+// authentication against the Vault server fails.
+func New(opts component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:           opts,
+		log:            opts.Logger,
+		stateChangedCh: make(chan struct{}, 1),
+	}
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the remote.vault component, managing the lifetime of the retrieved
+// secret and renewing/rereading it as necessary.
+func (c *Component) Run(ctx context.Context) error {
+	// TODO(rfratto): manage lifecycle of auth tokens
```

Comment:
Two things to do here:

* Re-read the secret:
  * On the configured refresh interval 
* If we need more than token auth:
  * Renew the auth token using [LifetimeWatcher](https://pkg.go.dev/github.com/hashicorp/vault/api#LifetimeWatcher), until it's no longer renewable; after which, get a new auth token.